### PR TITLE
New OpenThread CLI Arduino Library for ESP32-C6 and ESP32-H2

### PR DESCRIPTION
**Pull request by SuGlider**
## Description of Change
Adds the OpenThread CLI Arduino Library. 
It has some examples.

Only available for the C6 and H2.

## Tests scenarios
Examples tested on C6 and H2.

### CODE REVIEW

**Comment by SuGlider** [comment]
@lucasssvaz - Is it possible that the CI failures are related to using cached files that don't have the necessary file structure?

File: libraries/OpenThread/examples/COAP/coap_lamp/ci.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+    "targets": {
+      "esp32": false
+      "esp32c2": false
+      "esp32c3": false
+      "esp32s2": false
+      "esp32s3": false
+    }
+  }
```
Comment (P-R-O-C-H-Y) [inline libraries/OpenThread/examples/COAP/coap_lamp/ci.json:9]: @SuGlider CI is failing due to missing commas in the ci.json
```suggestion
{
    "targets": {
      "esp32": false,
      "esp32c2": false,
      "esp32c3": false,
      "esp32s2": false,
      "esp32s3": false
    }
  }
```

**Comment by P-R-O-C-H-Y** [comment]
@SuGlider Also you have to fix the Arduino as IDF component as CI is failing for C6 and H2.
```
In file included from /__w/arduino-esp32/arduino-esp32/components/arduino-esp32/libraries/OpenThread/src/OThreadCLI.cpp:1:
/__w/arduino-esp32/arduino-esp32/components/arduino-esp32/libraries/OpenThread/src/OThreadCLI.h:5:10: fatal error: esp_openthread.h: No such file or directory
    5 | #include "esp_openthread.h"
      |          ^~~~~~~~~~~~~~~~~~
Missing "esp_openthread.h" file name found in the following component(s): openthread(/opt/esp/idf/components/openthread/include/esp_openthread.h). 
Maybe one of the components needs to add the missing header directory to INCLUDE_DIRS of idf_component_register call in CMakeLists.txt.
```

File: libraries/OpenThread/examples/COAP/coap_lamp/ci.json (hunk 1 above)
Comment (SuGlider) [reply libraries/OpenThread/examples/COAP/coap_lamp/ci.json]: Thank @P-R-O-C-H-Y !!

**Comment by SuGlider** [comment]
> @SuGlider Also you have to fix the Arduino as IDF component as CI is failing for C6 and H2.
> 
> ```
> In file included from /__w/arduino-esp32/arduino-esp32/components/arduino-esp32/libraries/OpenThread/src/OThreadCLI.cpp:1:
> /__w/arduino-esp32/arduino-esp32/components/arduino-esp32/libraries/OpenThread/src/OThreadCLI.h:5:10: fatal error: esp_openthread.h: No such file or directory
>     5 | #include "esp_openthread.h"
>       |          ^~~~~~~~~~~~~~~~~~
> Missing "esp_openthread.h" file name found in the following component(s): openthread(/opt/esp/idf/components/openthread/include/esp_openthread.h). 
> Maybe one of the components needs to add the missing header directory to INCLUDE_DIRS of idf_component_register call in CMakeLists.txt.
> ```

Openthread is already  an IDF component. 
https://github.com/espressif/esp-idf/tree/master/components/openthread
But there are 2 folders there that are external repositories and those should be cloned by CI in a recursive way.

When I use Arduino as Component, I clone IDF repository with all submodules, recursively and this error never happens.

**Comment by me-no-dev** [comment]
@SuGlider I suggest you guard the library files with `#if CONFIG_OPENTHREAD_ENABLED`. 
```cpp
#include "sdkconfig.h"
#if CONFIG_OPENTHREAD_ENABLED
#include "esp_openthread.h"
//...
#endif /* CONFIG_OPENTHREAD_ENABLED */
```

**Comment by me-no-dev** [comment]
> Openthread is already an IDF component.

but it's not enabled in the default CI build. That's it's purpose

**Comment by SuGlider** [comment]
@P-R-O-C-H-Y  - DangerJS Linter fails... it seems that this PR has too many commits... it that the problem?

**Comment by lucasssvaz** [comment]
@SuGlider The DangerJS failure is bc some of the wokwi tests failed

**Comment by SuGlider** [comment]
> @SuGlider The DangerJS failure is bc some of the wokwi tests failed https://github.com/espressif/arduino-esp32/runs/26578220915

But all passed, only DangerJS failed.

File: CMakeLists.txt [hunk 2]
```diff
@@ -158,7 +159,11 @@ set(ARDUINO_LIBRARY_LittleFS_SRCS libraries/LittleFS/src/LittleFS.cpp)
 
 set(ARDUINO_LIBRARY_NetBIOS_SRCS libraries/NetBIOS/src/NetBIOS.cpp)
 
-set(ARDUINO_LIBRARY_PPP_SRCS
+set(ARDUINO_LIBRARY_OpenThread_SRCS
+  libraries/OpenThread/src/OThreadCLI.cpp
+  libraries/OpenThread/src/OThreadCLI_Util.cpp)
+
+  set(ARDUINO_LIBRARY_PPP_SRCS
```
Comment (me-no-dev) [inline CMakeLists.txt:166]: extra TAB was added here

File: libraries/OpenThread/src/OThreadCLI.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  bool otStarted = false;
+public:
+  OpenThreadCLI();
+  ~OpenThreadCLI();
+  // returns true if OpenThread CLI is running
+  operator bool() const;
+
+  // starts a task to read/write otStream. Default prompt is "ot> ". Set it to NULL to make it invisible.
+  void startOpenThreadConsole(Stream& otStream, bool echoback = true, const char *prompt = "ot> ");
+  void stopOpenThreadConsole();
```
Comment (me-no-dev) [inline libraries/OpenThread/src/OThreadCLI.h:35]: how about `startConsole` and `endConsole`?

**Review by me-no-dev — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 2 above)
Comment (SuGlider) [reply CMakeLists.txt:166]: Done.

File: libraries/OpenThread/src/OThreadCLI.h (hunk 3 above)
Comment (SuGlider) [reply libraries/OpenThread/src/OThreadCLI.h:35]: ok.

Comment (SuGlider) [reply libraries/OpenThread/src/OThreadCLI.h]: Done.

**Review by me-no-dev — APPROVED** [review]

**Review by lucasssvaz — APPROVED** [review]

File: libraries/OpenThread/examples/COAP/coap_lamp/coap_lamp.ino [hunk 4]
```diff
@@ -7,19 +7,41 @@
 #define OT_COAP_RESOURCE_NAME "Lamp"
 
 const char *otSetupLeader[] = {
```
Comment (lucasssvaz) [inline libraries/OpenThread/examples/COAP/coap_lamp/coap_lamp.ino:9]:
```suggestion
const char *otSetupLeader[] = {
  // clang-format off
```

File: libraries/OpenThread/examples/COAP/coap_lamp/coap_lamp.ino [hunk 5]
```diff
@@ -7,19 +7,41 @@
...
+  // set the network key 
+  "dataset networkkey", OT_NETWORK_KEY,
+  // commit the dataset
+  "dataset", "commit active",
+  // -- network start
+  // start the interface
+  "ifconfig", "up",
+  // start the Thread network
+  "thread", "start"
 };
```
Comment (lucasssvaz) [inline libraries/OpenThread/examples/COAP/coap_lamp/coap_lamp.ino:33]:
```suggestion
  // clang-format on
};
```

File: libraries/OpenThread/examples/COAP/coap_lamp/coap_lamp.ino [hunk 6]
```diff
@@ -7,19 +7,41 @@
...
 
 const char *otCoapLamp[] = {
```
Comment (lucasssvaz) [inline libraries/OpenThread/examples/COAP/coap_lamp/coap_lamp.ino:35]:
```suggestion
const char *otCoapLamp[] = {
  // clang-format off
```

File: libraries/OpenThread/examples/COAP/coap_lamp/coap_lamp.ino [hunk 7]
```diff
@@ -7,19 +7,41 @@
...
-  // start and create a CoAP resource
-  "coap", "start", "coap resource", OT_COAP_RESOURCE_NAME, "coap set", "0"
+  // -- start and create a CoAP resource
+  // start CoAP as server
+  "coap", "start",
+  // create a CoAP resource
+  "coap resource", OT_COAP_RESOURCE_NAME,
+  // set the CoAP resource initial value
+  "coap set", "0"
 };
```
Comment (lucasssvaz) [inline libraries/OpenThread/examples/COAP/coap_lamp/coap_lamp.ino:45]:
```suggestion
  // clang-format on
};
```

File: libraries/OpenThread/examples/COAP/coap_switch/coap_switch.ino [hunk 8]
```diff
@@ -8,18 +8,32 @@
 #define OT_COAP_RESOURCE_NAME "Lamp"
 
 const char *otSetupChild[] = {
```
Comment (lucasssvaz) [inline libraries/OpenThread/examples/COAP/coap_switch/coap_switch.ino:10]:
```suggestion
const char *otSetupChild[] = {
  // clang-format off
```

File: libraries/OpenThread/examples/COAP/coap_switch/coap_switch.ino [hunk 9]
```diff
@@ -8,18 +8,32 @@
...
+  // set the network key 
+  "dataset networkkey", OT_NETWORK_KEY,
+  // commit the dataset
+  "dataset", "commit active",
+  // -- network start
+  // start the interface
+  "ifconfig", "up",
+  // start the Thread network
+  "thread", "start"
 };
```
Comment (lucasssvaz) [inline libraries/OpenThread/examples/COAP/coap_switch/coap_switch.ino:32]:
```suggestion
  // clang-format on
};
```

File: libraries/OpenThread/examples/COAP/coap_switch/coap_switch.ino [hunk 10]
```diff
@@ -8,18 +8,32 @@
...
 
 const char *otCoapSwitch[] = {
```
Comment (lucasssvaz) [inline libraries/OpenThread/examples/COAP/coap_switch/coap_switch.ino:34]:
```suggestion
const char *otCoapSwitch[] = {
  // clang-format off
```

File: libraries/OpenThread/examples/COAP/coap_switch/coap_switch.ino [hunk 11]
```diff
@@ -8,18 +8,32 @@
...
-  // start and create a CoAP resource
-  "coap",
-  "start",
+  // -- start CoAP as client
+  "coap", "start"
 };
```
Comment (lucasssvaz) [inline libraries/OpenThread/examples/COAP/coap_switch/coap_switch.ino:37]:
```suggestion
  // clang-format on
};
```

**Review by lucasssvaz — CHANGES_REQUESTED** [review]
To avoid putting everything in a single line

**Comment by SuGlider** [comment]
@lucasssvaz - I have added more commented lines in the array. I think that it should get all in the right place. Could youforce running the formatting CI job on top of the current PR just to check if it changes anything else?
Thanks.

**Review by lucasssvaz — APPROVED** [review]
@SuGlider Yeah, looks good.

**Comment by SuGlider** [comment]
@me-no-dev  - I guess that the PR is ready for merging. Please check if there is something else missing beside the Wokwi CI errors.

**Pull request merged**